fitbenchmarking/core/results_output.py

```diff
@@ -199,6 +202,171 @@ def preprocess_data(results: "list[FittingResult]"):
     return best_results, sorted_results
 
 
+def _handle_fallback_tags(results,
```

Comment:
I ran an edge case and it broke. I think it's because I've chosen to compare multiple hessians but have kept the jacobian as the default and in the dataset one problem has analytic jacobians whereas the other doesn't. 

I don't know what the best way to fix this function would be. I'll send you the checkpoint file so you don't have to run it unless you want to change something on the data generation side of fb.

ini file
```
[FITTING]
hes_method: scipy
            numdifftools

[MINIMIZERS]
ralfit: hybrid
```

command
```
fitbenchmarking -p fitbenchmarking/test_files/default_parsers_set/ -o test.ini -s ralfit -d -n 1
```

output
```
Unknown exception. Exiting.
Traceback (most recent call last):
  File "/home/alister/venvs/fitbenchmarking/bin/fitbenchmarking-cp", line 33, in <module>
    sys.exit(load_entry_point('FitBenchmarking', 'console_scripts', 'fitbenchmarking-cp')())
  File "/home/alister/fitbenchmarking/fitbenchmarking/cli/checkpoint_handler.py", line 126, in main
    generate_report(args.options_file,
  File "/home/alister/fitbenchmarking/fitbenchmarking/cli/exception_handler.py", line 29, in wrapped
    return f(*args, **kwargs)
  File "/home/alister/fitbenchmarking/fitbenchmarking/cli/checkpoint_handler.py", line 99, in generate_report
    directory = save_results(
  File "/home/alister/fitbenchmarking/fitbenchmarking/utils/write_files.py", line 29, in wrapper
    return function(*args, **kwargs)
  File "/home/alister/fitbenchmarking/fitbenchmarking/core/results_output.py", line 58, in save_results
    best_results, results_dict = preprocess_data(results)
  File "/home/alister/fitbenchmarking/fitbenchmarking/core/results_output.py", line 200, in preprocess_data
    best_results[r][c] = _process_best_results(cat)
  File "/home/alister/fitbenchmarking/fitbenchmarking/core/results_output.py", line 431, in _process_best_results
    if best.accuracy > result.accuracy:
AttributeError: 'NoneType' object has no attribute 'accuracy'
```

Reply:
This is a very interesting edge case find. It is worth adding to the tests for results_output(). 

I had one question about the following command. Are you able to run the equivalent of this command (with analytic jacobians) on the master branch ?

`fitbenchmarking -p fitbenchmarking/test_files/default_parsers_set/ -o test.ini -s ralfit -d -n 1`

I am loading the new `checkpoint.json` file and calling `preprocess_data()` _with the new functions on this branch commented out and undoing the changes between line 142 to 151._

sorted_results is still populated with `None` values. So, I wanted to confirm if this command was possible before. If yes, could you kindly share a picture of the output table. I just want to see how it should look.

Reply:
I have updated the logic of `_handle_fallback_tags()`. The updated logic is more compact.

**But there is a new issue with the ralfit edge case**

These are the fallback columns in the ralfit checkpoint.json file.

`['ralfit:hybrid:analytic:scipy 2-point',
   'ralfit:hybrid:analytic:numdifftools central', 
   'ralfit:hybrid:scipy 2-point:scipy 2-point', 
   'ralfit:hybrid:scipy 2-point:numdifftools central']`

The renaming convention that we discussed should rename both Jacobin and Hessian to `best_available` in this case. This is exactly what happens. All the tags are renamed as `ralfit:hybrid:best_available:best_available`. 

However, during the subsequent processing in `preprocess_data()`. sorted_results is created with a shape of 3x1 instead of 3x2. _And I think the results are mislabelled because of overwritting._

I have tried rethinking the renaming convention but I can't think of anything that would work for this edge case.

Reply:
**This is just for illustration**. The ralfit results without column renaming would have this structure. 

<img width="547" alt="image" src="https://github.com/fitbenchmarking/fitbenchmarking/assets/47083562/4b8b2ebf-4284-47b8-8686-37034508e2bc">

The alternative to renaming the columns would be if we handle the `None` values on the frontend. It might make the tables messier if there are a lot of fallbacks. But at least all the information would be displayed on the frontend.

Reply:
Can you try this command again with the latest commits ?

`fitbenchmarking -p fitbenchmarking/test_files/default_parsers_set/ -o test.ini -s ralfit -d -n 1`.

I updated the `_handle_fallback_tags()` again. **I don't think we should ever update both the jacobian and the hessian tags to `best_avaliable` at the same time**. This will lead to multiple results mapping to the same columns.

 I have added an additional condition to update just one tag (I have chosen to update the Jacobian tag for now).

Reply:
> These are the fallback columns in the ralfit checkpoint.json file.
> 
> `['ralfit:hybrid:analytic:scipy 2-point', 'ralfit:hybrid:analytic:numdifftools central', 'ralfit:hybrid:scipy 2-point:scipy 2-point', 'ralfit:hybrid:scipy 2-point:numdifftools central']`
> 
> The renaming convention that we discussed should rename both Jacobin and Hessian to `best_available` in this case. This is exactly what happens. All the tags are renamed as `ralfit:hybrid:best_available:best_available`.
> 

In this case (which is why I chose it to test), the hessians are not `best_available`. They are user chosen hessians.

I think the correct result would be 
`['ralfit:hybrid:best_available:scipy 2-point', 'ralfit:hybrid:best_available:numdifftools central', 'ralfit:hybrid:best_available:scipy 2-point', 'ralfit:hybrid:best_available:numdifftools central']`


> I updated the _handle_fallback_tags() again. I don't think we should ever update both the jacobian and the hessian tags to best_avaliable at the same time. This will lead to multiple results mapping to the same columns.

I think we can update both at the same time, we just have to be careful to make sure that it is the case that the used the fallback.


I'll have a look at the update this afternoon. We could always add 2 bools to the fitting result if it's using a jacobian fallback or hessian fallback?

Reply:
With the latest changes, this edge case should be processed correctly and should not raise an error. I will also add the tests for this edge case today.

However, If there are other edge cases, then I can have a look at the fallback bools in `FittingResults`. I considered it. I also added them but removed them later because I thought the logic would be more complicated and would require iterating over the list of results multiple times. I abandoned the approach because I could not think of an easy way of doing it at that time but it can be revisited if problems with edge cases persist.

_What I meant about not updating the hessian and jacobian tags at the same time was_ - I don't think that we should every allow a column tag to be renamed as `s1:m1:best_avaliable:best_avaliable`. Because in this case we would end up mapping the results of 4 partially filled columns into 1 column. We would overwrite and possibly mislabel the results.